model/project_parser_db.go

```diff
@@ -61,56 +65,101 @@ func ParserProjectUpdateOne(query interface{}, update interface{}) error {
+func setAllFieldsUpdate(pp *ParserProject) interface{} {
+	return bson.M{
+		"$set": bson.M{
+			ParserProjectEnabledKey:           pp.Enabled,
+			ParserProjectStepbackKey:          pp.Stepback,
+			ParserProjectPreErrorFailsTaskKey: pp.PreErrorFailsTask,
+			ParserProjectBatchTimeKey:         pp.BatchTime,
+			ParserProjectOwnerKey:             pp.Owner,
+			ParserProjectRepoKey:              pp.Repo,
+			ParserProjectRepoKindKey:          pp.RepoKind,
+			ParserProjectRemotePathKey:        pp.RemotePath,
+			ParserProjectBranchKey:            pp.Branch,
+			ParserProjectIdentifierKey:        pp.Identifier,
+			ParserProjectDisplayNameKey:       pp.DisplayName,
+			ParserProjectCommandTypeKey:       pp.CommandType,
+			ParserProjectIgnoreKey:            pp.Ignore,
+			ParserProjectPreKey:               pp.Pre,
+			ParserProjectPostKey:              pp.Post,
+			ParserProjectTimeoutKey:           pp.Timeout,
+			ParserProjectCallbackTimeoutKey:   pp.CallbackTimeout,
+			ParserProjectModulesKey:           pp.Modules,
+			ParserProjectBuildVariantsKey:     pp.BuildVariants,
+			ParserProjectFunctionsKey:         pp.Functions,
+			ParserProjectTaskGroupsKey:        pp.TaskGroups,
+			ParserProjectTasksKey:             pp.Tasks,
+			ParserProjectExecTimeoutSecsKey:   pp.ExecTimeoutSecs,
+			ParserProjectLoggersKey:           pp.Loggers,
+			//ParserProjectAxesKey:              pp.Axes,
+			ParserProjectConfigNumberKey: pp.ConfigUpdateNumber,
+		},
+	}
+}
+
+// TryInsert suppresses the error of inserting if it's a duplicate key error and attempts to
+// upsert if config number matches
+func (pp *ParserProject) TryUpsert() error {
+	err := pp.Insert()
+	if err == nil || !strings.Contains(err.Error(), "duplicate key error") {
+		return errors.Wrapf(err, "database error inserting parser project")
+	}
+
+	q := bson.M{
+		ParserProjectConfigNumberKey: bson.M{"$exists": false},
+		ParserProjectConfigNumberKey: pp.ConfigUpdateNumber,
+	}
+	err = ParserProjectUpdateOne(q, setAllFieldsUpdate(pp))
+	if err == nil || adb.ResultsNotFound(err) {
+		return nil
+	}
+
+	return errors.Wrapf(err, "database error updating parser project")
+}
+
+// UpsertWithConfigNumber inserts project if it DNE. Otherwise, updates if the
+// existing config number is less than or equal to the new config number.
+// If shouldEqual, only update if the config update number matches.
+func (pp *ParserProject) UpsertWithConfigNumber(num int, shouldEqual bool) error {
 	if pp.Id == "" {
 		return errors.New("no version ID given")
 	}
-
+	pp.ConfigUpdateNumber = num + 1
 	found, err := ParserProjectFindOneById(pp.Id)
 	if err != nil {
 		return errors.Wrapf(err, "error finding parser project '%s'", pp.Id)
 	}
+	// the document doesn't exist yet, so just insert it
 	if found == nil {
-		pp.ConfigUpdateNumber = num
-		return errors.Wrap(pp.Insert(), "error inserting parser project")
+		if err = pp.Insert(); err == nil {
+			return nil
+		}
+		grip.Debug(message.WrapError(err, message.Fields{
```

Comment:
Shouldn't we error if the insert failed?

Reply:
No, and I dropped a comment to try and explain: 
// if the insert didn't succeed, likely this ID has now already been inserted, so we should overwrite it

This is actually the exact same approach as TryUpsert() except we don't case on the error string, we just assume in an error that we should Update instead of Upsert. I can make these consistent if we decide on an approach.